reference/python.md

```diff
@@ -17,16 +20,23 @@ This will leave behind a `Pulumi.yaml` file, containing some minimal metadata ab
 
 > **Note:** Although the template uses a very simple package structure, by placing `__main__.py` in the root directory, Pulumi fully supports [properly modularized Python programs](http://docs.python-guide.org/en/latest/writing/structure/) and `setup.py` files.  This is important if you ever decide to turn your Pulumi program into a library.
 
+> **Note:** Pulumi currently expects the `python` executable to refer to a Python of version 3.6 or above. This is
+> usually not the case when running outside of a virtual environment. To work around around this and explicitly ask
+> Pulumi to run your program using `python3` you can set the `PULUMI_PYTHON_CMD` environment variable to `python3`. This
+> will be addressed in future versions of Pulumi.
+
 ## Using Pulumi PyPI Packages {#pypi-packages}
 
-It is not required, but we recommend using [`virtualenv`](http://docs.python-guide.org/en/latest/dev/virtualenvs/) for repeatable installations that work no matter your operating system.
+It is not required, but we recommend using [`pipenv`](https://pipenv.readthedocs.io/en/latest/) for repeatable installations that work no matter your operating system.
 
 ```
-$ virtualenv .
-$ . ./bin/activate
+$ pipenv --python 3
+$ pipenv install
```

Comment:
When I follow these instructions myself - I see this:

```
$ pipenv --python 3
Creating a virtualenv for this project...
Pipfile: /Users/luke/go/src/github.com/pulumi/examples/aws-py-stepfunctions/Pipfile
Using /usr/local/bin/python3 (3.7.0) to create virtualenv...
⠋Running virtualenv with interpreter /usr/local/bin/python3
Using base prefix '/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7'
/usr/local/Cellar/pipenv/2018.7.1/libexec/lib/python3.7/site-packages/virtualenv.py:1041: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
New python executable in /Users/luke/.local/share/virtualenvs/aws-py-stepfunctions-W_WOZN1B/bin/python3.7
Also creating executable in /Users/luke/.local/share/virtualenvs/aws-py-stepfunctions-W_WOZN1B/bin/python
Installing setuptools, pip, wheel...done.
Setting project for aws-py-stepfunctions-W_WOZN1B to /Users/luke/go/src/github.com/pulumi/examples/aws-py-stepfunctions

Virtualenv location: /Users/luke/.local/share/virtualenvs/aws-py-stepfunctions-W_WOZN1B
requirements.txt found, instead of Pipfile! Converting...
Warning: Your Pipfile now contains pinned versions, if your requirements.txt did.
We recommend updating your Pipfile to specify the "*" version, instead.

luke:~/go/src/github.com/pulumi/examples/aws-py-stepfunctions (lukehoban/python3)
$ pipenv install
Pipfile.lock not found, creating...
Locking [dev-packages] dependencies...
Locking [packages] dependencies...
lib/python3.7/site-packages/pipenv/utils.py", line 402, in resolve_deps
    req_dir=req_dir
  File "/usr/local/Cellar/pipenv/2018.7.1/libexec/lib/python3.7/site-packages/pipenv/utils.py", line 250, in actually_resolve_deps
    req = Requirement.from_line(dep)
  File "/usr/local/Cellar/pipenv/2018.7.1/libexec/lib/python3.7/site-packages/pipenv/vendor/requirementslib/models/requirements.py", line 704, in from_line
    line, extras = _strip_extras(line)
TypeError: 'module' object is not callable
```

Two questions:
1. Is the error here related to our packages somehow?
2. If we are going to recommend moving to `pipenv` should we be moving from `requirements.txt` to `Pipfile` in examples/templates/etc.?

Note that I have not had any problems using `virtualenv`.

Reply:
> Is the error here related to our packages somehow?
> TypeError: 'module' object is not callable

No, that's a bug in pipenv. Arg... I guess pipenv is still quite buggy. Pipenv 2018.7.1 is quite old, though, they might have fixed that issue already.

Virtualenv is certainly the more stable option, and if you're having trouble with it I'm sure you won't be the last, so I should probably leave the virtualenv instructions in place.